server/plugin.go

```diff
@@ -0,0 +1,373 @@
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+)
+
+type Plugin struct {
+	plugin.MattermostPlugin
+
+	// configurationLock synchronizes access to the configuration.
+	configurationLock sync.RWMutex
+
+	// configuration is the active plugin configuration. Consult getConfiguration and
+	// setConfiguration for usage.
+	configuration *configuration
+
+	TeamID    string
+	ChannelID string
+	BotUserID string
+}
+
+const (
+	SNS_ICON_URL = "https://cdn2.iconfinder.com/data/icons/amazon-aws-stencils/100/App_Services_copy_Amazon_SNS-512.png"
+	SNS_USERNAME = "AWS SNS Bot"
+)
+
+func (p *Plugin) OnActivate() error {
+	configuration := p.getConfiguration()
+
+	if err := p.IsValid(configuration); err != nil {
+		return err
+	}
+
+	split := strings.Split(p.configuration.TeamChannel, ",")
+	teamSplit := split[0]
+	channelSplit := split[1]
+
+	team, err := p.API.GetTeamByName(teamSplit)
+	if err != nil {
+		return err
+	}
+	p.TeamID = team.Id
+
+	user, err := p.API.GetUserByUsername(p.configuration.Username)
+	if err != nil {
+		p.API.LogError(err.Error())
+		return fmt.Errorf("Unable to find user with configured username: %v", p.configuration.Username)
+	}
+	p.BotUserID = user.Id
+
+	channel, err := p.API.GetChannelByName(team.Id, channelSplit, false)
+	if err != nil && err.StatusCode == http.StatusNotFound {
+		channelToCreate := &model.Channel{
+			Name:        channelSplit,
+			DisplayName: channelSplit,
+			Type:        model.CHANNEL_OPEN,
+			TeamId:      p.TeamID,
+			CreatorId:   p.BotUserID,
+		}
+
+		newChannel, errChannel := p.API.CreateChannel(channelToCreate)
+		if err != nil {
+			return errChannel
+		}
+		p.ChannelID = newChannel.Id
+	} else if err != nil {
+		return err
+	} else {
+		p.ChannelID = channel.Id
+	}
+
+	return nil
+}
+
+func (p *Plugin) IsValid(configuration *configuration) error {
+	if configuration.TeamChannel == "" {
+		return fmt.Errorf("Must set a Team and a Channel.")
+	}
+
+	if configuration.AllowedUserIds == "" {
+		return fmt.Errorf("Must set at least one User.")
+	}
+
+	return nil
+}
+
+func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
+	if err := p.checkToken(r); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		p.API.LogError("AWSSNS TOKEN INVALID")
+		return
+	}
+	snsMessageType := r.Header.Get("x-amz-sns-message-type")
+	if snsMessageType == "" {
+		p.handleAction(w, r)
+
+	} else {
+		switch snsMessageType {
+		case "SubscriptionConfirmation":
+			p.handleSubscriptionConfirmation(r.Body)
+			break
+		case "Notification":
+			p.API.LogDebug("AWSSNS HandleNotification")
+			p.handleNotification(r.Body)
+			break
+
+		case "UnsubscribeConfirmation":
+			p.handleUnsubscribeConfirmation(r.Body)
+			break
+		default:
+			break
+		}
+	}
+
+	return
+
+}
+
+func (p *Plugin) checkToken(r *http.Request) error {
+	token := r.URL.Query().Get("token")
```

Comment:
prefer https://golang.org/pkg/net/http/#Request.FormValue

Reply:
0/5 not worth being a function